### PR TITLE
ncm-network: Add OpenVSwitch support for network interface

### PR DESCRIPTION
**Pull request by Pansanel**

### CODE REVIEW

**Comment by ned21** [comment]
OK to test

File: ncm-network/src/main/pan/components/network/core-schema.pan [hunk 1]
```diff
@@ -125,6 +125,9 @@ type structure_interface = {
     "stp" ? boolean # enable/disable stp on bridge (true: STP=on)
     "delay" ? long # brctl setfd DELAY
     "bridging_opts" ? structure_bridging_options
+
+    "devicetype" ? string with match(SELF, '^(ovs)$')
+    "ovs_bridge" ? string with exists ("/system/network/interfaces/" + SELF)
 };
```
Comment (stdweird) [inline ncm-network/src/main/pan/components/network/core-schema.pan]: the logic in the component that log errors should be part of the schema.

**Comment by stdweird** [comment]
@Pansanel thx, this saves @alvarosimon some work :smile:

**Comment by alvarosimon** [comment]
@Pansanel thanks! yes indeed :smile:

File: ncm-network/src/main/pan/components/network/core-schema.pan (hunk 1 above)
Comment (Pansanel) [inline ncm-network/src/main/pan/components/network/core-schema.pan]: I cannot find an example with two conditions checks in schema. Do you have some examples?

Comment (stdweird) [inline ncm-network/src/main/pan/components/network/core-schema.pan]: here's one for ncm-network https://github.com/quattor/configuration-modules-core/pull/420

**Comment by Pansanel** [comment]
I did some changes according to the comments. In addition, as described in /usr/share/doc/openvswitch-2.1.2/README.RHEL, devicetype is now setup to 'ovs' by default when interface type is OVSPort or OVSBridge.

**Comment by ned21** [comment]
Did the buildbot ignore my OK to test? :(

**Comment by stdweird** [comment]
ok to test

**Comment by stdweird** [comment]
@ned21 odd, i'd have to check if it is case sensitive though.

**Comment by stdweird** [comment]
OK to test

**Comment by stdweird** [comment]
@ned21 it is case sensitive. i modified jenkins so you can now use OK too

**Comment by stdweird** [comment]
OK to test

File: ncm-network/src/main/pan/components/network/core-schema.pan [hunk 2]
```diff
@@ -100,7 +100,7 @@ type structure_interface = {
     "driver"  ? string
     "bootproto" ? string
     "onboot" ? string
-    "type"    ? string with match(SELF, '^(Ethernet|Bridge|Tap|xDSL)$')
+    "type"    ? string with match(SELF, '^(Ethernet|Bridge|Tap|xDSL|OVSBridge|OVSPort)$')
```
Comment (alvarosimon) [inline ncm-network/src/main/pan/components/network/core-schema.pan]: @Pansanel maybe you should also include OVS_EXTRA to configure extra OVS options for bridges and ports, in some cases this option is quite useful (to set the VLAN trunk for example), I found this from ovs github doc: 
http://git.openvswitch.org/cgi-bin/gitweb.cgi?p=openvswitch;a=blob_plain;f=rhel/README.RHEL;hb=HEAD

Comment (Pansanel) [inline ncm-network/src/main/pan/components/network/core-schema.pan]: Hi,
I may add OVS_EXTRA, but I cannot test it yet (we are not using it at IPHC).
Jerome

**Comment by jouvin** [comment]
Any good reason to postpone merging this PR. We may add support for `OVS_EXTRA` as part of another issue/PR as Jerome says he cannot test it if adding it...

**Comment by Pansanel** [comment]
We can merge this PR. I will add support for OVS_EXTRA for the next release (15.4)

**Comment by jrha** [comment]
If possible I would like a test for this new functionality.

**Comment by jrha** [comment]
@Pansanel you can add the `OVS_EXTRA` support to this PR, if you could provide a test as well that would be awesome.

**Comment by Pansanel** [comment]
Ok, I will do it.

File: ncm-network/src/test/perl/ovs.t [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+my $cfg = get_config_for_profile('ovs');
+my $cmp = NCM::Component::network->new('network');
+
+is($cmp->Configure($cfg), 1, "Component runs correctly with a test profile");
+
+my $bfh = get_file($cmp->gen_backup_filename("/etc/sysconfig/network-scripts/ifcfg-br-ex").NCM::Component::network::FAILED_SUFFIX);
+isa_ok($bfh,"CAF::FileWriter","This is a CAF::FileWriter network/ifcfg-br-ex file written");
+
+like($bfh, qr/TYPE=OVSBridge/m, "set type for the OVS bridge");
```
Comment (stdweird) [inline ncm-network/src/test/perl/ovs.t]: can you use stricter regexp (`^TYPE=OVSBridge$`)?

**Comment by stdweird** [comment]
@Pansanel small remark wrt using stricter regexp, looks fine otherwise. thanks a lot!

File: ncm-network/src/main/perl/network.pm [hunk 4]
```diff
@@ -303,6 +303,31 @@ sub Configure {
...
+                    $text .= "OVS_BRIDGE='$net{$iface}{ovs_bridge}'\n";
+                }
+                if ($net{$iface}{'ovs_opts'}) {
+                    $text .= "OVS_OPTIONS='$net{$iface}{ovs_opts}'\n";
+                }
+                if ($net{$iface}{'ovs_extra'}) {
+                    $text .= "OVS_EXTRA='$net{$iface}{ovs_extra}'\n";
+                }
+                if ($net{$iface}{'bond_ifaces'}) {
+                    $text .= "BOND_IFACES=".join(' ',$net{$iface}{bond_ifaces})."\n";
```
Comment (stdweird) [inline ncm-network/src/main/perl/network.pm]: @Pansanel i'm doing a review for the whole component, but this needs single quotes right?
(pls don't push a fix, i'll fix it in the other PR)

Comment (Pansanel) [inline ncm-network/src/main/perl/network.pm]: For some parameters, like BOND_IFACES or OVS_OPTIONS, quotes or double quotes seems to be mandatory. For DEVICETYPE and TYPE, it is not the case. What is wrong with single quotes?

Comment (stdweird) [inline ncm-network/src/main/perl/network.pm]: these so-called config files are sourced in a bash environment, so proper quoting around whitespace is required. there's nothing worng with single quotes, but there are misisng here :smile: 
fixed in https://github.com/quattor/configuration-modules-core/pull/461/files#diff-90bc6b2e2a105ae5c62daf55ef95b2a7R283

**Comment by jrha** [comment]
LGTM thanks!
@stdweird if I understand you are addressing your own comments in #461 and this can be merged.

**Comment by stdweird** [comment]
@jrha yes, so i'll leave it to @Pansanel to make comments to the other PR :smile:

**Pull request merged**